### PR TITLE
Workload Bursting - Integration tests

### PR DESCRIPTION
**Pull request by meln1k**
Workload Bursting - Integration tests

Summary: Integration tests for Workload Bursting: 
* App is created in home region
* App is created in selected region
* App is created in selected region and zone 

JIRA issues: MARATHON_EE-1671

### CODE REVIEW

File: src/main/scala/mesosphere/marathon/util/FaultDomain.scala [hunk 1]
```diff
@@ -0,0 +1,4 @@
+package mesosphere.marathon
+package util
```
Comment (ichernetsky) [inline src/main/scala/mesosphere/marathon/util/FaultDomain.scala]: I think it should be placed somewhere else, not in `util`. What do you think?

File: src/test/scala/mesosphere/marathon/integration/setup/MesosTest.scala [hunk 2]
```diff
@@ -37,18 +37,41 @@ case class MesosCluster(
...
   system: ActorSystem,
     mat: Materializer,
     ctx: ExecutionContext,
     scheduler: Scheduler) extends AutoCloseable {
   require(quorumSize > 0 && quorumSize <= numMasters)
+  require(agentsFaultDomains.isEmpty || agentsFaultDomains.size == numSlaves)
 
   lazy val masters = 0.until(numMasters).map { i =>
+    val faultDomainJson = if (mastersFaultDomains.nonEmpty && mastersFaultDomains(i).nonEmpty) {
+      val fd = mastersFaultDomains(i).get
```
Comment (ichernetsky) [inline src/test/scala/mesosphere/marathon/integration/setup/MesosTest.scala]: Please rename `fd` to `faultDomain` or something else which is more readable.

File: src/test/scala/mesosphere/marathon/integration/setup/MesosTest.scala [hunk 3]
```diff
@@ -57,9 +80,35 @@ case class MesosCluster(
     // to it's marathon, leading to multiple tasks (from different IT suits) trying to use the same port!
     // First-come-first-served task will bind successfully where the others will fail leading to a lot inconsistency and
     // flakiness in tests.
+
+    val (additionalAttributes: Map[String, Option[String]], faultDomainJson) = if (agentsFaultDomains.nonEmpty && agentsFaultDomains(i).nonEmpty) {
+      val fd = agentsFaultDomains(i).get
```
Comment (ichernetsky) [inline src/test/scala/mesosphere/marathon/integration/setup/MesosTest.scala]: The same here.

File: src/main/scala/mesosphere/marathon/MarathonScheduler.scala [hunk 4]
```diff
@@ -30,7 +30,7 @@ class MarathonScheduler(
   private[this] val log = LoggerFactory.getLogger(getClass.getName)
 
   private var lastMesosMasterVersion: Option[SemanticVersion] = Option.empty
-  @volatile private[this] var homeRegion: Option[String] = Option.empty
+  @volatile private[this] var homeFaultDomain: Option[FaultDomain] = Option.empty
```
Comment (kensipe) [inline src/main/scala/mesosphere/marathon/MarathonScheduler.scala]: Really do NOT like the name change to FaultDomain... this is not a fault domain.   there is no described fault domain... that is left to the user.   homeRegion makes sense... homeDomain... defaultDomain

File: src/main/scala/mesosphere/marathon/MarathonScheduler.scala [hunk 5]
```diff
@@ -148,22 +148,25 @@ class MarathonScheduler(
     }
   }
 
-  protected def updateHomeRegion(masterInfo: MasterInfo): Unit = {
+  protected def updateHomeFaultDomain(masterInfo: MasterInfo): Unit = {
     if (masterInfo.hasDomain && masterInfo.getDomain.hasFaultDomain) {
-      homeRegion = Some(masterInfo.getDomain.getFaultDomain.getRegion.getName)
+      homeFaultDomain = Some(FaultDomain(
+        masterInfo.getDomain.getFaultDomain.getRegion.getName,
+        masterInfo.getDomain.getFaultDomain.getZone.getName
```
Comment (kensipe) [inline src/main/scala/mesosphere/marathon/MarathonScheduler.scala]: I see where you get the name (mesos)... still don't like it

File: src/main/scala/mesosphere/marathon/util/FaultDomain.scala (hunk 1 above)
Comment (kensipe) [reply src/main/scala/mesosphere/marathon/util/FaultDomain.scala]: completely agree with ivan! 
We should have a domain package... or a constraint package.. refactor the constraints into constraint package and add this.  
Util is a waste land... and should be lower level technical functions... not high order business functions / classes.

File: src/main/scala/mesosphere/marathon/MarathonScheduler.scala (hunk 5 above)
Comment (kensipe) [reply src/main/scala/mesosphere/marathon/MarathonScheduler.scala]: The issue as I see it with fault domain... given a default cluster (3 racks of 3 nodes each).   Dan could have 1 cluster with 3 fault domains (each rack).  and still may want another region to support work load bursting.     What is the "homeFaultDomain"?

File: src/main/scala/mesosphere/marathon/MarathonScheduler.scala (hunk 4 above)
Comment (alenkacz) [reply src/main/scala/mesosphere/marathon/MarathonScheduler.scala]: +1

Comment (alenkacz) [reply src/main/scala/mesosphere/marathon/MarathonScheduler.scala]: I would vote for keeping home region...

Comment (jeschkies) [reply src/main/scala/mesosphere/marathon/MarathonScheduler.scala]: We do call it `FaultDomain` though, don't we? @meichstedt what do the other teams call these? `Region` or `FaultDomain`?

File: src/main/scala/mesosphere/marathon/core/launcher/InstanceOpFactory.scala [hunk 6]
```diff
@@ -29,7 +30,7 @@ object InstanceOpFactory {
     * @param additionalLaunches the number of additional launches that has been requested
     */
   case class Request(runSpec: RunSpec, offer: Mesos.Offer, instanceMap: Map[Instance.Id, Instance],
-      additionalLaunches: Int) {
+      additionalLaunches: Int, homeFaultDomain: Option[FaultDomain]) {
```
Comment (jeschkies) [inline src/main/scala/mesosphere/marathon/core/launcher/InstanceOpFactory.scala]: Why not simply `faultDomain`? What am I missing?

File: src/main/scala/mesosphere/marathon/core/matcher/base/util/ActorOfferMatcher.scala [hunk 7]
```diff
@@ -29,13 +30,6 @@ class ActorOfferMatcher(actorRef: ActorRef, override val precedenceFor: Option[P
 
   override def toString: String = s"ActorOfferMatcher($actorRef)"
 
-  override def isInterestedIn(offer: Offer) = {
```
Comment (jeschkies) [inline src/main/scala/mesosphere/marathon/core/matcher/base/util/ActorOfferMatcher.scala]: Why was this removed?

File: src/main/scala/mesosphere/marathon/util/FaultDomain.scala (hunk 1 above)
Comment (jeschkies) [reply src/main/scala/mesosphere/marathon/util/FaultDomain.scala]: Hm, this could be part of `state` or `mesosphere.marathon.core.instance`.

File: src/main/scala/mesosphere/mesos/ResourceMatcher.scala [hunk 8]
```diff
@@ -133,7 +134,7 @@ object ResourceMatcher extends StrictLogging {
     * the reservation.
     */
   def matchResources(offer: Offer, runSpec: RunSpec, knownInstances: => Seq[Instance],
-    selector: ResourceSelector, conf: MatcherConf, schedulerPlugins: Seq[SchedulerPlugin] = Seq.empty)(implicit clock: Clock): ResourceMatchResponse = {
+    selector: ResourceSelector, conf: MatcherConf, schedulerPlugins: Seq[SchedulerPlugin] = Seq.empty, homeFaultDomain: Option[FaultDomain])(implicit clock: Clock): ResourceMatchResponse = {
```
Comment (jeschkies) [inline src/main/scala/mesosphere/mesos/ResourceMatcher.scala]: Ok, this is a little confusing. I thought this diff is about testing the workload bursting. However, it seems we introduce the feature here. Could you shed a little light on this? 🙂

File: src/main/scala/mesosphere/marathon/MarathonScheduler.scala (hunk 4 above)
Comment (meln1k) [reply src/main/scala/mesosphere/marathon/MarathonScheduler.scala]: @kensipe LocalDomain?

File: src/main/scala/mesosphere/mesos/ResourceMatcher.scala (hunk 8 above)
Comment (meln1k) [reply src/main/scala/mesosphere/mesos/ResourceMatcher.scala]: I agree. I moved all feature-related code to another PR: https://github.com/mesosphere/marathon/pull/5673

File: src/main/scala/mesosphere/marathon/MarathonScheduler.scala (hunk 4 above)
Comment (alenkacz) [reply src/main/scala/mesosphere/marathon/MarathonScheduler.scala]: Here it is called home region https://docs.google.com/document/d/11qJtAR8aT7ZHrUjBwsDtet9FVeIGre7GQdppF9pbcdE/edit#heading=h.v9bqn0eo8mhe and I kind of like that :-)

File: src/main/scala/mesosphere/mesos/ResourceMatcher.scala [hunk 9]
```diff
@@ -176,6 +176,12 @@ object ResourceMatcher extends StrictLogging {
 
     def portsMatchOpt: Option[PortsMatch] = PortsMatcher(runSpec, offer, selector).portsMatch
 
+    val meetsFaultDomainRequirements: Boolean = {
```
Comment (alenkacz) [inline src/main/scala/mesosphere/mesos/ResourceMatcher.scala]: why was this moved from offermatcher to resourcematcher?

File: src/main/scala/mesosphere/marathon/state/FaultDomain.scala [hunk 10]
```diff
@@ -3,6 +3,10 @@ package state
 
 case class FaultDomain(region: Region, zone: Zone)
 
-case class Region(value: String) extends AnyVal
+case class Region(value: String) extends AnyVal {
```
Comment (zen-dog) [inline src/main/scala/mesosphere/marathon/state/FaultDomain.scala]: 👍 to value classes

File: src/test/scala/mesosphere/marathon/integration/RemoteRegionOffersIntegrationTest.scala [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+  val remoteZone2 = Zone("remote_zone2")
+
+  override def mastersFaultDomains = Seq(Some(FaultDomain(region = homeRegion, zone = homeZone)))
+
+  override def agentsFaultDomains = Seq(
+    Some(FaultDomain(region = remoteRegion, zone = remoteZone1)),
+    Some(FaultDomain(region = remoteRegion, zone = remoteZone2)),
+    Some(FaultDomain(region = homeRegion, zone = homeZone)))
+
+  before(cleanUp())
```
Comment (zen-dog) [inline src/test/scala/mesosphere/marathon/integration/RemoteRegionOffersIntegrationTest.scala]: why do you need to `cleanUp` between tests? We try to avoid that and make every test a _snow flake_ independent of the rest of the cluster. Ideally `cleanUp` should only be called once at the end of the test.

File: src/test/scala/mesosphere/marathon/integration/RemoteRegionOffersIntegrationTest.scala [hunk 12]
```diff
@@ -0,0 +1,110 @@
...
+
+  def appId(suffix: Option[String] = None): PathId = testBasePath / s"app-${suffix.getOrElse(UUID.randomUUID)}"
```
Comment (zen-dog) [inline src/test/scala/mesosphere/marathon/integration/RemoteRegionOffersIntegrationTest.scala]: Let's give every test a meaningful and "unique" name. Makes later debugging soo much easier. Suffix here should be mandatory.

File: src/test/scala/mesosphere/marathon/integration/RemoteRegionOffersIntegrationTest.scala [hunk 13]
```diff
@@ -0,0 +1,110 @@
...
+
+      Then("The app is created in the default region")
+      result should be(Created)
+      extractDeploymentIds(result)
+      waitForDeployment(result)
+      waitForTasks(app.id.toPath, 1)
+      val slaveId = marathon.tasks(applicationId).value.head.slaveId.get
+      val agentRegion = mesos.state.value.agents.find(_.id == slaveId).get.attributes.attributes("fault_domain_region")
+
+      agentRegion match {
```
Comment (zen-dog) [inline src/test/scala/mesosphere/marathon/integration/RemoteRegionOffersIntegrationTest.scala]: `inside(agentRegion)` may be? Otherwise will a no match fail the test?

File: src/test/scala/mesosphere/marathon/integration/RemoteRegionOffersIntegrationTest.scala [hunk 14]
```diff
@@ -0,0 +1,110 @@
...
+        case ITResourceStringValue(value) => value shouldEqual remoteRegion.value
+      }
+      agentZone match {
+        case ITResourceStringValue(value) => value shouldEqual remoteZone2.value
+      }
+    }
+  }
+
+}
```
Comment (zen-dog) [inline src/test/scala/mesosphere/marathon/integration/RemoteRegionOffersIntegrationTest.scala]: Looks like you can get away with 2 agents since you don't use `remoteZone1`?

File: src/test/scala/mesosphere/marathon/integration/setup/MesosTest.scala [hunk 15]
```diff
@@ -57,9 +81,35 @@ case class MesosCluster(
...
+          |          "name": "${faultDomain.region}"
+          |        },
+          |      "zone":
+          |        {
+          |          "name": "${faultDomain.zone}"
+          |        }
+          |    }
+          |}
+        """.stripMargin
+      Map("fault_domain_region" -> Some(faultDomain.region.value), "fault_domain_zone" -> Some(faultDomain.zone.value)) -> Some(faultDomainJson)
```
Comment (zen-dog) [inline src/test/scala/mesosphere/marathon/integration/setup/MesosTest.scala]: This is a hard line to read. And `additionalAttributes` are actually `faultDomainAttributes`

File: src/test/scala/mesosphere/marathon/integration/RemoteRegionOffersIntegrationTest.scala (hunk 14 above)
Comment (meln1k) [reply src/test/scala/mesosphere/marathon/integration/RemoteRegionOffersIntegrationTest.scala]: We still have to see that we don't place anything on the other agent

File: src/test/scala/mesosphere/marathon/integration/facades/MesosFacade.scala [hunk 16]
```diff
@@ -61,7 +61,7 @@ object MesosFacade {
     def apply(vals: (String, Any)*): ITResources = {
       val resources: Map[String, ITResourceValue] = vals.map {
         case (id, value: Double) => id -> ITResourceScalarValue(value)
-        case (id, portsString: String) => id -> ITResourcePortValue(portsString)
+        case (id, portsString: String) => id -> ITResourceStringValue(portsString)
```
Comment (jeschkies) [inline src/test/scala/mesosphere/marathon/integration/facades/MesosFacade.scala]: Nit: should this be `value` instead of `postString` then?

File: src/test/scala/mesosphere/marathon/integration/setup/MesosTest.scala [hunk 17]
```diff
@@ -57,9 +81,40 @@ case class MesosCluster(
     // to it's marathon, leading to multiple tasks (from different IT suits) trying to use the same port!
     // First-come-first-served task will bind successfully where the others will fail leading to a lot inconsistency and
     // flakiness in tests.
+
+    val (faultDomainAttributes: Map[String, Option[String]], faultDomainJson) = if (agentsFaultDomains.nonEmpty && agentsFaultDomains(i).nonEmpty) {
+      val faultDomain = agentsFaultDomains(i).get
+      val faultDomainJson = s"""
```
Comment (jeschkies) [inline src/test/scala/mesosphere/marathon/integration/setup/MesosTest.scala]: Hm, I have a hard time understand the type of `faultDomainJson`. 

What is

```scala
         |}
        """.stripMargin
      (
        Map(
        "fault_domain_region" -> Some(faultDomain.region.value),
        "fault_domain_zone" -> Some(faultDomain.zone.value)),
        Some(faultDomainJson)
      )
```

doing?

File: src/test/scala/mesosphere/marathon/integration/setup/MesosTest.scala [hunk 18]
```diff
@@ -57,9 +81,40 @@ case class MesosCluster(
...
+        "fault_domain_region" -> Some(faultDomain.region.value),
+        "fault_domain_zone" -> Some(faultDomain.zone.value)),
+        Some(faultDomainJson)
+      )
+    } else Map.empty -> None
+
+    // uniquely identify each agent node, useful for constraint matching
+    val attributes: Map[String, Option[String]] = Map("node" -> Some(i.toString)) ++ faultDomainAttributes
+
+    val renderedAttributes = attributes.map { case (key, maybeVal) => s"$key${maybeVal.map(v => s":$v").getOrElse("")}" }.mkString(";")
```
Comment (jeschkies) [inline src/test/scala/mesosphere/marathon/integration/setup/MesosTest.scala]: Is this a string? Please add a type.

File: src/test/scala/mesosphere/marathon/integration/setup/MesosTest.scala [hunk 19]
```diff
@@ -57,9 +81,40 @@ case class MesosCluster(
...
+          |    }
+          |}
+        """.stripMargin
+      (
+        Map(
+        "fault_domain_region" -> Some(faultDomain.region.value),
+        "fault_domain_zone" -> Some(faultDomain.zone.value)),
+        Some(faultDomainJson)
+      )
+    } else Map.empty -> None
```
Comment (jeschkies) [inline src/test/scala/mesosphere/marathon/integration/setup/MesosTest.scala]: Ok, if I'm not mistaken the the fault domain is just a simple Map. So why not create this? What is the benefit of the JSON here? I'm confused.™

Comment (meln1k) [reply src/test/scala/mesosphere/marathon/integration/setup/MesosTest.scala]: We have to pass json as option argument for mesos fault domain. More here: http://mesos.apache.org/documentation/latest/configuration/master-and-agent/

File: src/test/scala/mesosphere/marathon/integration/setup/MesosTest.scala (hunk 17 above)
Comment (meln1k) [reply src/test/scala/mesosphere/marathon/integration/setup/MesosTest.scala]: It provides agent attributes and cmd option for enabling fault domain in mesos

File: src/test/scala/mesosphere/marathon/integration/facades/MesosFacade.scala (hunk 16 above)
Comment (meln1k) [reply src/test/scala/mesosphere/marathon/integration/facades/MesosFacade.scala]: It should!

File: src/test/scala/mesosphere/marathon/integration/RemoteRegionOffersIntegrationTest.scala [hunk 20]
```diff
@@ -0,0 +1,118 @@
...
+@IntegrationTest
+class RemoteRegionOffersIntegrationTest extends AkkaIntegrationTest with EmbeddedMarathonTest with Inside {
+
+  override lazy val mesosNumMasters = 1
+  override lazy val mesosNumSlaves = 3
+
+  val homeRegion = Region("home_region")
+  val homeZone = Zone("home_zone")
+
+  val remoteRegion = Region("remote_region")
```
[the remark this answers is not in the document]
Comment (meln1k) [reply src/test/scala/mesosphere/marathon/integration/RemoteRegionOffersIntegrationTest.scala]: Not sure if that brings more value, but I can do that.

File: src/test/scala/mesosphere/marathon/integration/RemoteRegionOffersIntegrationTest.scala [hunk 21]
```diff
@@ -0,0 +1,118 @@
...
+    "Launch an instance of the app in the default region if region is not specified" in {
+      val applicationId = appId("must-be-placed-in-home-region")
+      val app = appProxy(applicationId, "v1", instances = 1, healthCheck = None)
+
+      When("The app is deployed without specifying region")
+      val result = marathon.createAppV2(app)
+
+      Then("The app is created in the default region")
+      result should be(Created)
+      extractDeploymentIds(result)
```
[the remark this answers is not in the document]
Comment (meln1k) [reply src/test/scala/mesosphere/marathon/integration/RemoteRegionOffersIntegrationTest.scala]: It's doing nothing :) good catch

File: src/test/scala/mesosphere/marathon/integration/RemoteRegionOffersIntegrationTest.scala (hunk 14 above)
Comment (zen-dog) [reply src/test/scala/mesosphere/marathon/integration/RemoteRegionOffersIntegrationTest.scala]: But where do you actually check it?

Comment (meln1k) [reply src/test/scala/mesosphere/marathon/integration/RemoteRegionOffersIntegrationTest.scala]:
```scala
inside(agentZone) {
        case ITResourceStringValue(value) => value shouldEqual remoteZone2.value
      }
```
We check that task was launched on the agent with a zone2 and not other agent.

File: src/test/scala/mesosphere/marathon/integration/RemoteRegionOffersIntegrationTest.scala (hunk 20 above)
[the remark this answers is not in the document]
Comment (jeschkies) [reply src/test/scala/mesosphere/marathon/integration/RemoteRegionOffersIntegrationTest.scala]: It organizes the code a little bit 🙂

File: src/test/scala/mesosphere/marathon/integration/RemoteRegionOffersIntegrationTest.scala [hunk 22]
```diff
@@ -0,0 +1,111 @@
...
+  class Fixture {
+    val homeRegion = Region("home_region")
+    val homeZone = Zone("home_zone")
+
+    val remoteRegion = Region("remote_region")
+    val remoteZone1 = Zone("remote_zone1")
+    val remoteZone2 = Zone("remote_zone2")
+  }
+
+  val fixture = new Fixture
```
Comment (jeschkies) [inline src/test/scala/mesosphere/marathon/integration/RemoteRegionOffersIntegrationTest.scala]: Please create the fixture in each test case. My idea is to encapsulate all fixtures.

File: src/test/scala/mesosphere/marathon/integration/facades/MesosFormats.scala [hunk 23]
```diff
@@ -12,20 +12,20 @@ object MesosFormats {
...
+  implicit lazy val ITResourcePortValueFormat: Format[ITResourceStringValue] = Format(
+    Reads.of[String].map(ITResourceStringValue(_)),
     Writes(portValue => JsString(portValue.portString))
   )
 
   implicit lazy val ITResourceValueFormat: Format[ITResourceValue] = Format(
     Reads[ITResourceValue] {
       case JsNumber(value) => JsSuccess(ITResourceScalarValue(value.toDouble))
-      case JsString(portsString) => JsSuccess(ITResourcePortValue(portsString))
+      case JsString(portsString) => JsSuccess(ITResourceStringValue(portsString))
```
Comment (jeschkies) [inline src/test/scala/mesosphere/marathon/integration/facades/MesosFormats.scala]: Nit: is `portsString` still appropriate here?

File: src/test/scala/mesosphere/marathon/integration/facades/MesosFormats.scala [hunk 24]
```diff
@@ -12,20 +12,20 @@ object MesosFormats {
...
       case _ => JsError("expected string or number")
     },
     Writes[ITResourceValue] {
       case ITResourceScalarValue(value) => JsNumber(value)
-      case ITResourcePortValue(portsString) => JsString(portsString)
+      case ITResourceStringValue(portsString) => JsString(portsString)
```
Comment (jeschkies) [inline src/test/scala/mesosphere/marathon/integration/facades/MesosFormats.scala]: Nit: see above.

File: src/test/scala/mesosphere/marathon/integration/RemoteRegionOffersIntegrationTest.scala (hunk 22 above)
Comment (meln1k) [reply src/test/scala/mesosphere/marathon/integration/RemoteRegionOffersIntegrationTest.scala]: Could you please elaborate on that?

Comment (jeschkies) [reply src/test/scala/mesosphere/marathon/integration/RemoteRegionOffersIntegrationTest.scala]: I just saw that you do

```
"Launch an instance of the app in the default region if region is not specified" in new Fixture { ... }
```

So why is this still required? Because of 

```
override def mastersFaultDomains = Seq(Some(FaultDomain(region = fixture.homeRegion, zone = fixture.homeZone)))
```
?

Hm, this does not seems very clean.

File: src/test/scala/mesosphere/marathon/integration/RemoteRegionOffersIntegrationTest.scala [hunk 25]
```diff
@@ -0,0 +1,111 @@
...
+
+  override def agentsFaultDomains = Seq(
+    Some(FaultDomain(region = fixture.remoteRegion, zone = fixture.remoteZone1)),
+    Some(FaultDomain(region = fixture.remoteRegion, zone = fixture.remoteZone2)),
+    Some(FaultDomain(region = fixture.homeRegion, zone = fixture.homeZone)))
+
+  def appId(suffix: String): PathId = testBasePath / s"app-${suffix}"
+
+  "Region Aware marathon" must {
+    "Launch an instance of the app in the default region if region is not specified" in new Fixture {
```
Comment (jeschkies) [inline src/test/scala/mesosphere/marathon/integration/RemoteRegionOffersIntegrationTest.scala]: [Nit]: I'd avoid this style of fixture creation and use `val f = new Fixture` and `f.homeRegion`. This way it's way easier to read where something is coming from. Imagine someone has now access to IntelliJ and has to be able to understand the code.
